### PR TITLE
net: dial retries

### LINKED ISSUE #706: net: dial retries
In case dial fails in an upstream Forwarder it's ineffective.
We should add dial retry with tiny exponential backoff.

### PR DESCRIPTION
**Pull request by mmatczuk**
Fixes #706

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ linters:
     - gomnd
     - gomoddirectives
     - inamedparam
+    - intrange
```
Comment (Choraden) [inline .golangci.yml:26]: Why not use range?

File: net.go [hunk 2]
```diff
@@ -155,6 +171,37 @@ func (d *Dialer) DialContext(ctx context.Context, network, address string) (net.
...
+
+		conn, err := dial(ctx, network, address)
+		if err == nil {
+			return conn, nil
+		}
+		lastErr = err
+
+		if conn != nil {
+			conn.Close()
+		}
```
Comment (Choraden) [inline net.go:200]: if err != nil, conn is nil. That's the rule we were following before this patch.

Comment (Choraden) [reply net.go:200]: Maybe it can stay. Better safe than sorry.

File: net.go [hunk 3]
```diff
@@ -79,6 +69,18 @@ func DefaultDialConfig() *DialConfig {
 	return &DialConfig{
 		DialTimeout:     25 * time.Second,
 		KeepAliveConfig: defaultKeepAliveConfig(),
+		Retry:           3,
+		RetryBackoff:    1 * time.Second,
+	}
```
Comment (Choraden) [inline net.go]: This does not belong here.

File: net.go (hunk 2 above)
Comment (mmatczuk) [reply net.go:200]: That's the rule overall but there is nothing that governs that rule.

**Review by Choraden — APPROVED** [review]

**Pull request merged**